Card identify ai, indexdb, hash functions, dialog search image

1vcian opened the pull request:
## Pokémon Card Detection Project Update

After several attempts with limited results, I believe I've finally reached the end — at least, I hope so! 🤞


https://github.com/user-attachments/assets/4c290665-177c-494f-8285-dfeb56832ef5




I trained an **Oriented Bounding Box Detection** model on 10k screenshots (auto-generated using a script) to teach it to recognize cards in a generic way. This will allow the model to be used without needing retraining in the future. Once the AI was trained, I implemented it on the client-side using **TensorFlow.js**. I spent quite a bit of time understanding the output structure, but eventually managed to get bounding boxes as output. (Note: I had to create many supporting functions to make everything run smoothly.)

Once the cards are detected (I only select them if the AI is more than 90% confident that it’s a card), I implemented several functions to enhance the detection process:

- **Percentual Hashing**: I implemented it using colors for more accurate collision detection, as opposed to the standard hashing methods.
- **IndexedDB**: I use IndexedDB to store hashes in the database when they are missing or when new cards are added to the `public/images` folder. The first time the app is opened, the hashes are generated and saved to IndexedDB, and on subsequent visits, they are retrieved from there.
- Various **support functions** were also developed to streamline the workflow.

Finally, I built the component that detects cards from the images, which includes comparison logic, cropping the bounding boxes detected by the AI, calculating the hash, and creating a list of possible matches based on the Hamming distance. It returns the best match along with four possible alternatives. Below is a screenshot showing the result.

It’s been a really rewarding project to work on, and I plan to create a personal, more detailed repository with additional information about this feature
Images
-  Images Upload
<img width="1023" alt="image" src="https://github.com/user-attachments/assets/d8b2825e-ea1e-415c-81a6-542d809ca73d" />
-  Results (all selected when shown)##
<img width="1049" alt="image" src="https://github.com/user-attachments/assets/16e79cbf-118b-40d3-978e-8baf3aedd21b" />
-  Show Edit Matches
<img width="284" alt="image" src="https://github.com/user-attachments/assets/4ae1e563-9467-4ac1-a4ae-ced967671044" />
(You can select another card from the top if the algorithm fails. If your card isn't there, sorry!)

The selected card count increments by that amount, updating the number of selected cards.

Use case scenario: You open a pack, take a screenshot, and upload it to the app to update all the new cards you’ve got.

<img width="213" alt="image" src="https://github.com/user-attachments/assets/36ba9a27-8118-41eb-bba8-5ad9f352befc" />
<img width="208" alt="image" src="https://github.com/user-attachments/assets/0eb0c51f-fbb3-4336-95e5-99afae019e96" />

marcelpanse commented on downloadImagesCards.py:
This file can be deleted. There is a javascript version now for both the scraper and downloading the images in /scripts/.

marcelpanse commented on frontend/src/services/PokemonCardDetectionServices.tsx:
```diff
@@ -0,0 +1,239 @@
+  private async fileToImage(imageFile: File): Promise<HTMLImageElement> {
+    return new Promise<HTMLImageElement>((resolve, reject) => {
+      const img = new Image()
+      img.src = URL.createObjectURL(imageFile)
```
CodeQL detected some vulnerabilities, could you apply these?

marcelpanse commented on frontend/src/components/PokemonCardDetectorComponent.tsx:
```diff
@@ -0,0 +1,585 @@
+        await hashStorageService.initDB()
+        const storedHashCount = await hashStorageService.getHashCount()
+
+        if (storedHashCount !== uniqueCards.length) {
+          console.log('Generating hashes for all cards...')
+
+          const batchSize = 80
+          const totalCards = uniqueCards.length
+          const allHashes = []
+
+          for (let i = 0; i < totalCards; i += batchSize) {
+            const batch = uniqueCards.slice(i, i + batchSize)
+
+            // Process one batch
+            const batchPromises = batch.map(async (card) => {
+              try {
+                const hash = await hashingService.calculatePerceptualHash(`/images/${card.image?.split('/').at(-1)}`)
+                return { id: card.card_id, hash }
+              } catch (error) {
+                console.error(`Error generating hash for card ${card.card_id}:`, error)
+                return null
+              }
+            })
```
Here if we add a new card to the promo pack, it will regenerate all the hashes of all cards. Instead you could check if the hash already exists and then skip it to save some time.

marcelpanse commented on frontend/src/components/PokemonCardDetectorComponent.tsx:
```diff
@@ -0,0 +1,585 @@
+        URL.revokeObjectURL(imageUrl)
+        resolve(extractedCards)
+      }
+      image.src = imageUrl
```
Another CodeQL vulnerability to fix

marcelpanse commented:
@1vcian could you also update the branch with the latest from main and you have to delete the lockfile and regenerate it with pnpm install and commit that to fix the build here.

marcelpanse commented:
I have some small UI requests, but it's probably easier if I just do it myself after this is merged :-)

1vcian commented:
@marcelpanse Sorry, but there are many files. I hope it's not too complicated to read and make the changes you want to apply. If you need any clarifications on my code, feel free to ask

1vcian commented:
> @1vcian I pushed a couple of small changes on your branch:
> - promoted the scan button to the main header since it's an awesome feature :)
> - i'm hiding the scan button while it's still generating the hashes because it's no use already showing that if you can't use it. We'll generate the hashes in the background and once ready we'll show the button
> - I think this feature will mainly be used from mobile where you play the app, make a screenshot and upload. But it didn't work on mobile because mainly because of the Scroll box. I fixed the mobile issues. Also removed the alert because I think it's pretty self-explanatory and it took too much space. 
>  
> Final question:
> 1. after uploading my screenshot it took quite a while before showing the cards (like 30 seconds?). Is it possible to show a progress bar here, or isn't there any "progress" to show. If not, maybe a spinner and a text that explains it can take up to a minute?
> 

Strange, I only tried from my PC, and it was instant. It could be a compatibility issue between TensorFlow.js and mobile. I’ll check it tomorrow.
do I need to do anything to test it quickly on my phone?

marcelpanse commented:
It was on my laptop, it seems to be instant now. Maybe it was the initial download of the models?

1vcian commented:
> It was on my laptop, it seems to be instant now. Maybe it was the initial download of the models?

Yeah, that could definitely be the reason! Honestly, I had a bit of a hard time at the beginning figuring out the tensor output, so I kind of hit pause without even realizing it. But if everything’s running smoothly now, when do you think you’ll release the update? And how’s it looking so far? Have you tried it with different inputs? I’m super curious to hear your thoughts on the whole process and the algorithms I implemented on the feature

marcelpanse commented:
@1vcian I did some more testing and moved some logic around. Now you'll always see the scan button, and if you click it it will show you a loader and progress while loading both the model and pre-processing the cards. The reason I moved it is because it was slowing down the site a lot the first time you opened the page and that is very bad for the lighthouse scores. Now its more clean what is happening. I'll push this live now! :-)
